rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -96,6 +96,47 @@ NodeGraph::get_topic_names_and_types(bool no_demangle) const
   return topics_and_types;
 }
 
+std::map<std::string, std::vector<std::string>>
+NodeGraph::get_action_names_and_types(rclcpp::Node action_node) const
+{
+  rcl_names_and_types_t action_names_and_types = rcl_get_zero_initialized_names_and_types();
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  auto ret = rcl_action_get_names_and_types(
+    node_base_->get_rcl_node_handle(),
+    &allocator,
+    &action_names_and_types);
+  if (ret != RCL_RET_OK) {
+    auto error_msg = std::string("failed to get action names and types: ") +
+      rcl_get_error_string().str;
+    rcl_reset_error();
+    if (rcl_names_and_types_fini(&action_names_and_types) != RCL_RET_OK) {
+      error_msg += std::string(", failed also to cleanup action names and types, leaking memory: ") +
+        rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    throw std::runtime_error(error_msg);
+  }
+
+  std::map<std::string, std::vector<std::string>> actions_and_types;
+  for (size_t i = 0; i < action_names_and_types.names.size; ++i) {
+    std::string action_name = action_names_and_types.names.data[i];
+    for (size_t j = 0; j < action_names_and_types.types[i].size; ++j) {
+      actions_and_types[action_name].emplace_back(action_names_and_types.types[i].data[j]);
+    }
+  }
+
+  ret = rcl_names_and_types_fini(&action_names_and_types);
+  if (ret != RCL_RET_OK) {
+    // *INDENT-OFF*
+    throw std::runtime_error(
```

Comment:
I think we should stick with throw_from_rcl_error to keep consistency because there are multiple other instances where `throw_from_rcl_error` is utilized for checking the reliability of `rcl`